src/xdg.rs

```diff
@@ -179,27 +181,30 @@ impl IconLoader {
     ///
     /// This will check all paths for available icons and store them for cheap
     /// lookup.
-    fn new() -> Self {
+    fn new(data_dirs: &[PathBuf]) -> Self {
         let mut icons = HashMap::new();
 
         // Check all paths for icons.
         //
         // Since the `ICON_PATHS` is in reverse order of our priority, we can just
         // insert every new icon into `icons` and it will correctly return the
         // closest match.
-        for (path, extension) in ICON_PATHS {
-            let mut read_dir = fs::read_dir(path).ok();
-            let entries = read_dir.iter_mut().flatten().flatten();
-            let files = entries.filter(|e| e.file_type().map_or(false, |e| e.is_file()));
-
-            // Iterate over all files in the directory.
-            for file in files {
-                let file_name = file.file_name().to_string_lossy().to_string();
-
-                // Store icon paths with the correct extension.
-                let name = file_name.rsplit_once('.').filter(|(_, ext)| ext == extension);
-                if let Some((name, _)) = name {
-                    let _ = icons.insert(name.to_owned(), file.path());
+        for base_path in data_dirs {
+            for (path, extension) in ICON_PATHS.iter().map(|(p, e)| (base_path.join(p), e)) {
```

Comment:
Seems like we could do some iterator magic here to combine the two with flatten? I feel like it's unnecessary to have two levels of indentation here.

Reply:
I just did like you did in the other branch with the same situation. See applications.

Reply:
If you mean the `DesktopEntries::new` loop, then this might be a better implementation:

```diff
diff --git a/src/xdg.rs b/src/xdg.rs
index c6f3b37..19adfcb 100644
--- a/src/xdg.rs
+++ b/src/xdg.rs
@@ -54,40 +54,40 @@ impl DesktopEntries {
 
         // Find all desktop files in these directories, then look for their icons and
         // executables.
-        for dir_entry in dirs.iter().flat_map(|d| fs::read_dir(d.join("applications")).ok()) {
-            for desktop_file in dir_entry
-                .filter_map(|entry| entry.ok())
-                .filter(|entry| entry.file_type().map_or(false, |ft| ft.is_file()))
-                .filter(|entry| entry.file_name().to_string_lossy().ends_with(".desktop"))
-                .flat_map(|entry| fs::read_to_string(entry.path()).ok())
-            {
-                let mut icon = None;
-                let mut exec = None;
-                let mut name = None;
-
-                for line in desktop_file.lines() {
-                    if let Some(value) = line.strip_prefix("Name=") {
-                        name = Some(value.to_owned());
-                    } else if let Some(value) = line.strip_prefix("Icon=") {
-                        icon = desktop_entries.loader.load(value, icon_size).ok();
-                    } else if let Some(value) = line.strip_prefix("Exec=") {
-                        exec = value.split(' ').next().map(String::from);
-                    }
-
-                    if icon.is_some() && exec.is_some() && name.is_some() {
-                        break;
-                    }
+        let desktop_dirs = dirs.iter().flat_map(|d| fs::read_dir(d.join("applications")).ok());
+        for desktop_file in desktop_dirs
+            .flatten()
+            .flatten()
+            .filter(|entry| entry.file_type().map_or(false, |ft| ft.is_file()))
+            .filter(|entry| entry.file_name().to_string_lossy().ends_with(".desktop"))
+            .flat_map(|entry| fs::read_to_string(entry.path()))
+        {
+            let mut icon = None;
+            let mut exec = None;
+            let mut name = None;
+
+            for line in desktop_file.lines() {
+                if let Some(value) = line.strip_prefix("Name=") {
+                    name = Some(value.to_owned());
+                } else if let Some(value) = line.strip_prefix("Icon=") {
+                    icon = desktop_entries.loader.load(value, icon_size).ok();
+                } else if let Some(value) = line.strip_prefix("Exec=") {
+                    exec = value.split(' ').next().map(String::from);
                 }
 
-                if let Some((name, exec)) = name.zip(exec) {
-                    let icon = match icon {
-                        Some(icon) => Rc::new(icon),
-                        None => placeholder_icon.clone(),
-                    };
-
-                    desktop_entries.entries.push(DesktopEntry { icon, name, exec });
+                if icon.is_some() && exec.is_some() && name.is_some() {
+                    break;
                 }
             }
+
+            if let Some((name, exec)) = name.zip(exec) {
+                let icon = match icon {
+                    Some(icon) => Rc::new(icon),
+                    None => placeholder_icon.clone(),
+                };
+
+                desktop_entries.entries.push(DesktopEntry { icon, name, exec });
+            }
         }
 
         Ok(desktop_entries)

```

It's probably still an improvement I think, but your case makes even more sense to flatten down because you don't have a billion iterator steps.

Reply:
I'd take care of my case then. You could apply your patch separately, but don't forget to account for the symlinks.